### PR TITLE
Set gloo process group for FSDP with CPU offload

### PR DESCRIPTION
**Pull request by ebsmothers**
Addresses #1977.

As discussed in the issue (see [this comment](https://github.com/pytorch/torchtune/issues/1977#issuecomment-2512906526)), FSDP's implementation of gradient clipping uses [_NormPartial](https://github.com/pytorch/pytorch/blob/64d44a39a10a3e4b7abc85ffadd831ace3227460/torch/distributed/tensor/_ops/_math_ops.py#L56), which requires comms primitives (specifically `all_reduce`). This means that when running with CPU offloading we need to initialize the gloo process group to calculate the grad norm for DTensors on CPU. For simplicity this PR enables it whenever CPU offloading is enabled regardless of gradient clipping.

Test plan:

Added a test case for gradient clipping + CPU offload to `test_full_finetune_distributed.py`.

```
pytest -m integration_test tests/recipes/test_full_finetune_distributed.py -k 'test_loss'
...
========= 3 passed, 1 deselected in 43.54s ========
```

### CODE REVIEW

File: recipes/full_finetune_distributed.py [hunk 1]
```diff
@@ -946,11 +946,13 @@ def recipe_main(cfg: DictConfig) -> None:
             "Distributed finetune recipe should be run via a distributed launcher."
             "If using tune CLI, please specify --nnodes 1 and --nproc_per_node [num_gpus]"
         )
-    init_process_group(backend="gloo" if cfg.device == "cpu" else "nccl")
+    process_group = "gloo" if cfg.device == "cpu" else "nccl"
     if cfg.get("fsdp_cpu_offload", False):
         # Utilize all available CPU cores for intra-op parallelism. This provides ~2x
         # speed up when benchmarking fused AdamW on CPU
         training.set_torch_num_threads()
+        process_group = "cuda:nccl,cpu:gloo"
```
Comment (felipemello1) [inline recipes/full_finetune_distributed.py:954]: n00b question: why not just make this the default every time, instead of `"gloo" if cfg.device == "cpu" else "nccl"`

Comment (felipemello1) [reply recipes/full_finetune_distributed.py:954]: image

Comment (ebsmothers) [reply recipes/full_finetune_distributed.py:954]: I actually wasn't sure myself why we do this. Well [now I know](https://github.com/pytorch/torchtune/pull/482#discussion_r1529635443). So yes, I think we can take your suggestion here

**Review by felipemello1 — APPROVED** [review]
this seems to be low risk, since it was tested in the dcp already. Approving to unblock.
